// obigo-js-ui-rnbs/src/vue-install.js
import { current as theme } from './features/theme'
import focus from './features/focus'

import { install as eventsInstall } from './features/events'

import Button from './components/button'
import ButtonGroup from './components/button-group/button-group.vue'
import ButtonGroupItem from './components/button-group/button-group-item.vue'
import Checkbox from './components/checkbox/index.vue'
import Clock from './components/clock'
import Footer from './components/footer'
import GridList from './components/grid-list/GridList.vue'
import GridListItem from './components/grid-list/GridListItem.vue'
import Header from './components/header'
import List from './components/list/List.vue'
import ListItem from './components/list/ListItem.vue'
import Popup from './components/popup/popup.vue'
import Progress from './components/progress'
import Spinner from './components/spinner'

import FocusDirective from './directives/obg-focus'

function registerComponents (_Vue) { // eslint-disable-line
  _Vue.component('spinner', Spinner)

  ;[
    ['button', Button],
    ['button-group', ButtonGroup],
    ['button-group-item', ButtonGroupItem],
    ['checkbox', Checkbox],
    ['clock', Clock],
    ['footer', Footer],
    ['grid-list', GridList],
    ['grid-list-item', GridListItem],
    ['header', Header],
    ['list', List],
    ['list-item', ListItem],
    ['popup', Popup],
    ['progress', Progress],
    ['spinner', Spinner]
  ].forEach(c => {
    _Vue.component('obg-' + c[0], c[1])
  })
}

function registerDirectives (_Vue) {
  [
    ['obg-focus', FocusDirective]
  ].forEach(d => {
    _Vue.directive(d[0], d[1])
  })
}

export var Vue

export default function (_Vue) {
  if (this.installed) {
    console.warn('Obigo-js-ui already installed in Vue.')
    return
  }

  Vue = _Vue

  eventsInstall(_Vue)
  // registerComponents(_Vue)
  registerDirectives(_Vue)

  _Vue.prototype.$obigoUI = {
    theme,
    focus
  }
  _Vue.prototype.$theme = theme
  _Vue.prototype.$focus = focus

  // _Vue.prototype.$event = eventbus object
  // _Vue.prototype.$hardkey = { hardkey object}
}
