// obigo-js-ui-rnbs/src/features/hardkey.js
/**
 * Created by krinjadl on 2017-04-25.
 */

let code = {
  HARDKEY_TYPE_NONE: 1000, // unknown hardkey type
  HARDKEY_BUTTON_HOME: 1001, // home hardkey
  HARDKEY_BUTTON_BACK: 1002, // back hardkey
  HARDKEY_ROTARY_ROTATE: 2001, // rotary rotate hardkey
  HARDKEY_ROTARY_ENTER: 2002, // rotaty enter hardkey
  HARDKEY_ROTARY_LEFT: 2003, // rotary left hardkey
  HARDKEY_ROTARY_RIGHT: 2004, // rotary right hardkey
  HARDKEY_ROTARY_UP: 2005, // rotary up hardkey
  HARDKEY_ROTARY_DOWN: 2006 // rotary down hardkey
}

let mode = {
  HARDKEY_MODE_NONE: 0, // unknown hardkey mode
  HARDKEY_MODE_PRESS: 1, // hardkey press
  HARDKEY_MODE_LONG_PRESS: 2, // hardkey long press
  HARDKEY_MODE_RELEASE: 3, // hardkey release
  HARDKEY_MODE_LEFT: 4, // rotary hardkey rotate left
  HARDKEY_MODE_RIGHT: 5 // rotary hardkey rotate right
}

function getKeyCodes () {
  if (navigator.userAgent.indexOf('Obigo') < 0) {
    code = {
      HARDKEY_TYPE_NONE: 1000, // unknown hardkey type
      HARDKEY_BUTTON_HOME: 81, // home hardkey
      HARDKEY_BUTTON_BACK: 51, // back hardkey
      HARDKEY_ROTARY_ROTATE: 2001, // rotary rotate hardkey
      HARDKEY_ROTARY_ENTER: 2002, // rotaty enter hardkey
      HARDKEY_ROTARY_LEFT: 88, // rotary left hardkey
      HARDKEY_ROTARY_RIGHT: 82, // rotary right hardkey
      HARDKEY_ROTARY_UP: 53, // rotary up hardkey
      HARDKEY_ROTARY_DOWN: 54 // rotary down hardkey
    }
  }
  return {
    code: code,
    mode: mode
  }
}

export const hardkeyCode = getKeyCodes()
